### PR TITLE
avoid upcasting by assigning dtype to noise tensor

### PR DESCRIPTION
**Pull request by prathikr**
If we enable mixed precision training for unconditional image generation, the loss calculation `loss = F.mse_loss(model_output, noise)` upcasts the loss from float16 to float32 due to the fact that noise is always a float32 tensor resulting in the following error. This PR aims to fix this bug.

Run Command: `accelerate launch train_unconditional.py --dataset_name="huggan/flowers-102-categories" --resolution=64 --center_crop --random_flip --output_dir="ddpm-ema-flowers-64" --use_ema --train_batch_size=16 --num_epochs=10 --gradient_accumulation_steps=1 --learning_rate=1e-4 --lr_warmup_steps=500 --mixed_precision=fp16`

```
Traceback (most recent call last):
  File "train_unconditional.py", line 692, in <module>
    main(args)
  File "train_unconditional.py", line 594, in main
    accelerator.backward(loss)
  File "/opt/conda/envs/ptca/lib/python3.8/site-packages/accelerate/accelerator.py", line 1739, in backward
    self.deepspeed_engine_wrapped.backward(loss, **kwargs)
  File "/opt/conda/envs/ptca/lib/python3.8/site-packages/accelerate/utils/deepspeed.py", line 167, in backward
    self.engine.backward(loss, **kwargs)
  File "/opt/conda/envs/ptca/lib/python3.8/site-packages/deepspeed/utils/nvtx.py", line 15, in wrapped_fn
    ret_val = func(*args, **kwargs)
  File "/opt/conda/envs/ptca/lib/python3.8/site-packages/deepspeed/runtime/engine.py", line 1796, in backward
    self.optimizer.backward(loss, retain_graph=retain_graph)
  File "/opt/conda/envs/ptca/lib/python3.8/site-packages/deepspeed/runtime/zero/stage_1_and_2.py", line 1890, in backward
    self.loss_scaler.backward(loss.float(), retain_graph=retain_graph)
  File "/opt/conda/envs/ptca/lib/python3.8/site-packages/deepspeed/runtime/fp16/loss_scaler.py", line 62, in backward
    scaled_loss.backward(retain_graph=retain_graph)
  File "/opt/conda/envs/ptca/lib/python3.8/site-packages/torch/_tensor.py", line 488, in backward
    torch.autograd.backward(
  File "/opt/conda/envs/ptca/lib/python3.8/site-packages/torch/autograd/__init__.py", line 204, in backward
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
RuntimeError: Found dtype Float but expected Half
```

### CODE REVIEW

**Comment by prathikr** [comment]
@patrickvonplaten @anton-l could I please get a review on this? Thank you!

**Comment by patrickvonplaten** [comment]
Hey @prathikr,

Could you add a quick code snippet that reproduces the error for `train_unconditional.py`? Just a quick bash command that runs the script would be great

**Comment by prathikr** [comment]
@patrickvonplaten sure, here you go:

`accelerate launch train_unconditional.py --dataset_name="huggan/flowers-102-categories" --resolution=64 --center_crop --random_flip --output_dir="ddpm-ema-flowers-64" --use_ema --train_batch_size=16 --num_epochs=10 --gradient_accumulation_steps=1 --learning_rate=1e-4 --lr_warmup_steps=500 --mixed_precision=fp16`

**Comment by patrickvonplaten** [comment]
cc @sayakpaul @williamberman can you take a look?

**Review by sayakpaul — APPROVED** [review]
LGTM!

The failing tests are unrelated here.

File: examples/unconditional_image_generation/train_unconditional.py [hunk 1]
```diff
@@ -562,7 +562,9 @@ def transform_images(examples):
 
             clean_images = batch["input"]
             # Sample noise that we'll add to the images
-            noise = torch.randn(clean_images.shape).to(clean_images.device)
+            noise = torch.randn(
+                clean_images.shape, dtype=(torch.float32 if args.mixed_precision == "no" else torch.float16)
+            ).to(clean_images.device)
```
Comment (williamberman) [inline examples/unconditional_image_generation/train_unconditional.py:567]: Can you break these out as two different lines pls

Comment (prathikr) [reply examples/unconditional_image_generation/train_unconditional.py]: @williamberman done!

**Comment by patrickvonplaten** [comment]
Hey @prathikr, can you run `make style` here to make sure we don't get linting errors?

**Comment by prathikr** [comment]
@patrickvonplaten done

**Review by sayakpaul — APPROVED** [review]
Thank you!

**Comment by prathikr** [comment]
@sayakpaul can you please merge? I do not have permission to merge.

**Comment by sayakpaul** [comment]
Sorry for the delay!

**Pull request merged**